tasm-lib/src/arithmetic/u128/overflowing_add.rs

```diff
@@ -0,0 +1,289 @@
+            (u128::MAX - 1, 2),
+        ] {
+            snippet.prop_add(a, b);
+            snippet.prop_add(b, a);
+        }
+
+        for i in 0..128 {
+            let a = u128::MAX - ((1u128 << i) - 1);
+            let b = 1u128 << i;
+
+            // sanity check of test input values
+            let (wrapped_add, is_overflow) = a.overflowing_add(b);
+            assert!(is_overflow, "i = {i}. a = {a}, b = {b}");
+            assert!(wrapped_add.is_zero());
+
+            snippet.prop_add(b, a);
+        }
+    }
+
+    impl OverflowingAddU128 {
+        fn prop_add(&self, lhs: u128, rhs: u128) {
+            let init_stack = self.setup_init_stack(lhs, rhs);
+
+            let expected = {
+                let (sum, overflow) = lhs.overflowing_add(rhs);
+                let mut stack = self.init_stack_for_isolated_run();
+                push_encodable(&mut stack, &sum);
+                push_encodable(&mut stack, &overflow);
+                stack
+            };
+
+            test_rust_equivalence_given_complete_state(
+                &ShadowedClosure::new(OverflowingAddU128),
+                &init_stack,
+                &[],
+                &NonDeterminism::default(),
+                &None,
+                Some(&expected),
+            );
+        }
+
+        fn setup_init_stack(&self, lhs: u128, rhs: u128) -> Vec<BFieldElement> {
+            [
+                self.init_stack_for_isolated_run(),
+                lhs.encode().into_iter().rev().collect_vec(),
+                rhs.encode().into_iter().rev().collect_vec(),
+            ]
+            .concat()
+        }
+    }
+
+    impl Closure for OverflowingAddU128 {
+        fn rust_shadow(&self, stack: &mut Vec<BFieldElement>) {
+            fn to_u128(a: u32, b: u32, c: u32, d: u32) -> u128 {
+                a as u128
+                    + b as u128 * (1u128 << 32)
+                    + c as u128 * (1u128 << 64)
+                    + d as u128 * (1u128 << 96)
+            }
+
+            // top element on stack
+            let a0: u32 = stack.pop().unwrap().try_into().unwrap();
+            let b0: u32 = stack.pop().unwrap().try_into().unwrap();
+            let c0: u32 = stack.pop().unwrap().try_into().unwrap();
+            let d0: u32 = stack.pop().unwrap().try_into().unwrap();
+            let ab0 = to_u128(a0, b0, c0, d0);
+
+            // second element on stack
+            let a1: u32 = stack.pop().unwrap().try_into().unwrap();
+            let b1: u32 = stack.pop().unwrap().try_into().unwrap();
+            let c1: u32 = stack.pop().unwrap().try_into().unwrap();
+            let d1: u32 = stack.pop().unwrap().try_into().unwrap();
+            let ab1 = to_u128(a1, b1, c1, d1);
+
+            let (sum, overflow) = ab0.overflowing_add(ab1);
+
+            let mut res = sum.encode();
+            for _ in 0..res.len() {
+                stack.push(res.pop().unwrap());
+            }
+
+            stack.push(bfe!(overflow as u64));
+        }
```

Comment:
In [4d55540c](https://github.com/TritonVM/tasm-lib/commit/4d55540c17323b2a6875bcdd0d5eec1503e146a1#diff-1b74ceafc6be01656205edafa9086bbcf80924c31fcb8d31c35e82e83cb50d92R116), I add `pop_encodable`, the dual to `push_encodable`. This lets you remove essentially all the boilerplate from this rust shadowing:

```rust
fn rust_shadow(&self, stack: &mut Vec<BFieldElement>) {
    let left = pop_encodable::<u128>(stack);
    let right = pop_encodable(stack);
    let (sum, is_overflow) = left.overflowing_add(right);
    push_encodable(stack, &sum);
    push_encodable(stack, &is_overflow);
}
```